### PR TITLE
[codegen/python] fix deprecation warning on default values

### LINKED ISSUE #12546: Deprecated parameters cause warnings to be shown to the user whether or not the parameter is in use from Python programs
### What happened?

Marking MLC inputs (parameters) as deprecated causes the warnings to be shown to the user whether or not the parameters are actually in use.

### Expected Behavior

Deprecation warnings should only be shown to users when the user is actually utilizing one of the deprecated parameters.

### Steps to reproduce

Mark an optional MLC input (parameter) as deprecated, perform an update without setting the optional, deprecated parameter.

### Output of `pulumi about`

n/a

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by dixler**
# Description

Removes Resource constructor deprecation checks in favor of deprecation checks in the ResourceArgs constructor.

Checks in the ResourceArgs constructor which will be changed to `ResourceArgs.__configure__` in #13825 and compatible with dicts and objects that implement `__getitem__` `__setitem__`.

Merge https://github.com/pulumi/pulumi/pull/13825 after and update `__internal_init__` logic to use the `ResourceArgs.__configure__`

Fixes #12546 

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Review by Frassle — CHANGES_REQUESTED** [review]

File: pkg/codegen/testing/test/testdata/regress-py-12546/schema.json [hunk 1]
```diff
@@ -27,7 +27,6 @@
           "default": "medium"
         },
         "diameter": {
-          "deprecationMessage": "Dear future maintainer, if there are changes here, make sure that this is printed before the value is set to the default or else this will always print.",
```
[the remark this answers is not in the document]
Comment (dixler) [reply pkg/codegen/testing/test/testdata/regress-py-12546/schema.json:30]: There are 3 more above it. I initially added it to add variety with a prior approach, but I'll revert this change as it's vestigial to this change.

[the remark this answers is not in the document]
Comment (dixler) [reply pkg/codegen/testing/test/testdata/regress-py-12546/schema.json]: Addressed

File: pkg/codegen/testing/test/testdata/regress-py-12546/python-extras/tests/test_codegen.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+class MyMocks(pulumi.runtime.Mocks):
+    def call(self, args):
+        return {}
+    def new_resource(self, args):
+        """
+        new_resource mocks resource construction calls. This function should return the physical identifier and the output properties
+        for the resource being constructed.
+
+        :param MockResourceArgs args.
+        """
```
Comment (Frassle) [inline pkg/codegen/testing/test/testdata/regress-py-12546/python-extras/tests/test_codegen.py:43]:
```suggestion
```

No need for doc comments about the abstract method in test code.

File: pkg/codegen/testing/test/testdata/regress-py-12546/python-extras/tests/test_codegen.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        return 'foo', args.inputs
+
+
+@pulumi.runtime.test
+def test_func_with_default_value(my_mocks):
+    with patch("warnings.warn") as mock_warn:
+        pulumi_plant.tree.v1.RubberTree("my-tree", pulumi_plant.tree.v1.RubberTreeArgs())
```
Comment (Frassle) [inline pkg/codegen/testing/test/testdata/regress-py-12546/python-extras/tests/test_codegen.py:50]: Wasn't totally obvious what this was testing to start, I would add a doc comment to this method saying what we're doing "constructing a resource with deprecated but default values and checking the default values don't trigger a deprecation warning"

**Review by Frassle — APPROVED** [review]

**Pull request merged**